ESP32-IDF + ADC fails to compile for ESP32-S2

### The problem

The ESP32-S2 family has 13-bit ADCs (unlike the other ESP32 variants having 12-bit).
Unfortunately the bit width is hard coded in [ `adc_sensor.cpp`](https://github.com/esphome/esphome/blob/dev/esphome/components/adc/adc_sensor.cpp) ([here](https://github.com/esphome/esphome/blob/297824e2d75e568a8196ccb9e62d1f2e36e02f53/esphome/components/adc/adc_sensor.cpp#L26) and [here](https://github.com/esphome/esphome/blob/297824e2d75e568a8196ccb9e62d1f2e36e02f53/esphome/components/adc/adc_sensor.cpp#L33)).
For ESP32-S2 it should be `ADC_WIDTH_BIT_13` instead of `ADC_WIDTH_BIT_12`.

**Possible fix:**
The ESP-IDF has a Macro for the bit width in `ADC_WIDTH_BIT_DEFAULT` but that can't be used directly due to int->enum conversion. Therefore I redefined the macro with proper type conversion at the top of `adc_sensor.cpp`:
```
#undef ADC_WIDTH_BIT_DEFAULT
#define ADC_WIDTH_BIT_DEFAULT   ((adc_bits_width_t) ((int)ADC_WIDTH_MAX-1))
```

This fixes this issue:
https://github.com/esphome/feature-requests/issues/996#issuecomment-974689175

### Which version of ESPHome has the issue?

2022.1.2

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

 2021.12.10

### What platform are you using?

ESP32-IDF

### Board

esp32-s2-saola-1 (Ai-Thinker ESP-12K)

### Component causing the issue

adc

### Example YAML snippet

```yaml
esphome:
  name: esp32s2-01

esp32:
  board: esp32-s2-saola-1
  variant: esp32s2
  framework:
    type: esp-idf

sensor:
  - platform: adc
    pin: GPIO3
    name: "esp32s2 adc"
```


### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esp32s2-02.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing esp32s2-01 (board: esp32-s2-saola-1; framework: espidf; platform: platformio/espressif32 @ 3.3.2)
--------------------------------------------------------------------------------
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf 3.40300.0 (4.3.0) 
 - tool-cmake 3.16.4 
 - tool-ninja 1.7.1 
Reading CMake configuration...
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
No dependencies
Compiling .pioenvs/esp32s2-01/src/esphome/components/adc/adc_sensor.o
Compiling .pioenvs/esp32s2-01/src/main.o
src/esphome/components/adc/adc_sensor.cpp: In member function 'virtual void esphome::adc::ADCSensor::setup()':
src/esphome/components/adc/adc_sensor.cpp:27:21: error: 'ADC_WIDTH_BIT_12' was not declared in this scope
   adc1_config_width(ADC_WIDTH_BIT_12);
                     ^~~~~~~~~~~~~~~~
src/esphome/components/adc/adc_sensor.cpp:27:21: note: suggested alternative: 'ADC_WIDTH_BIT_13'
   adc1_config_width(ADC_WIDTH_BIT_12);
                     ^~~~~~~~~~~~~~~~
                     ADC_WIDTH_BIT_13
*** [.pioenvs/esp32s2-01/src/esphome/components/adc/adc_sensor.o] Error 1
========================= [FAILED] Took 35.13 seconds =========================
```


### Additional information

_No response_

--- a/esphome/components/adc/adc_sensor.cpp
+++ b/esphome/components/adc/adc_sensor.cpp
@@ -17,7 +17,9 @@ namespace adc {
 static const char *const TAG = "adc";
 // 13 bits for S3 / 12 bit for all other esp32 variants
 // create a const to avoid the repated cast to enum
-static const adc_bits_width_t ADC_WIDTH_MAX_SOC_BITS = static_cast<adc_bits_width_t>(ADC_WIDTH_BIT_DEFAULT);
+#ifdef USE_ESP32
+static const adc_bits_width_t ADC_WIDTH_MAX_SOC_BITS = static_cast<adc_bits_width_t>(ADC_WIDTH_MAX - 1);
+#endif
 
 void ADCSensor::setup() {
   ESP_LOGCONFIG(TAG, "Setting up ADC '%s'...", this->get_name().c_str());

--- a/esphome/components/adc/adc_sensor.cpp
+++ b/esphome/components/adc/adc_sensor.cpp
@@ -15,6 +15,9 @@ namespace esphome {
 namespace adc {
 
 static const char *const TAG = "adc";
+// 13 bits for S3 / 12 bit for all other esp32 variants
+// create a const to avoid the repated cast to enum
+static const adc_bits_width_t ADC_WIDTH_MAX_SOC_BITS = static_cast<adc_bits_width_t>(ADC_WIDTH_BIT_DEFAULT);
 
 void ADCSensor::setup() {
   ESP_LOGCONFIG(TAG, "Setting up ADC '%s'...", this->get_name().c_str());
@@ -23,14 +26,14 @@ void ADCSensor::setup() {
 #endif
 
 #ifdef USE_ESP32
-  adc1_config_width(ADC_WIDTH_BIT_12);
+  adc1_config_width(ADC_WIDTH_MAX_SOC_BITS);
   if (!autorange_) {
     adc1_config_channel_atten(channel_, attenuation_);
   }
 
   // load characteristics for each attenuation
   for (int i = 0; i < (int) ADC_ATTEN_MAX; i++) {
-    auto cal_value = esp_adc_cal_characterize(ADC_UNIT_1, (adc_atten_t) i, ADC_WIDTH_BIT_12,
+    auto cal_value = esp_adc_cal_characterize(ADC_UNIT_1, (adc_atten_t) i, ADC_WIDTH_MAX_SOC_BITS,
                                               1100,  // default vref
                                               &cal_characteristics_[i]);
     switch (cal_value) {

